File: training-data/php/integer_sqrt.php
<?php

function isqrt(int $n): int
{
    if ($n < 0) {
        throw new ValueError('negative');
    }
    if ($n < 2) {
        return $n;
    }
    $x = $n;
    $y = intdiv($x, 2);
    while ($y < $x) {
        $x = $y;
        $y = intdiv($x + intdiv($n, $x), 2);
    }
    return $x;
}

foreach ([0, 1, 2, 3, 15, 16, 17, 1_000_000, PHP_INT_MAX] as $n) {
    echo "isqrt($n) = ", isqrt($n), PHP_EOL;
}

File: training-data/python/integer_sqrt.py
def isqrt(n):
    if n < 0:
        raise ValueError("negative")
    lo, hi = 0, n
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if mid * mid <= n:
            lo = mid
        else:
            hi = mid - 1
    return lo


if __name__ == "__main__":
    import math
    for n in (0, 1, 15, 16, 17, 10 ** 20 + 12345):
        assert isqrt(n) == math.isqrt(n)
        print(n, isqrt(n))
